Remote Containers using SSH doesn't work in Windows

Related to microsoft/vscode-remote-release#1935

- I confirmed that I can successfully connect to the remote machine using SSH via command line
- There is no error in the entire UI, just loading progress bars when opening the Docker viewlet
- When running `Attach to Running Container`, nothing happens
- When I open devtools, this is what I see:

![image](https://user-images.githubusercontent.com/22350/70048648-14d0e400-15cb-11ea-946e-3ef839e581b1.png)

Additionally, I see theses messages if I open the vscode developer tools from Help:
`mainThreadExtensionService.ts:66 Error: All configured authentication methods failed`

--- a/src/utils/delay.ts
+++ b/src/utils/delay.ts
@@ -3,8 +3,23 @@
  *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-export async function delay(ms: number): Promise<void> {
-    return new Promise<void>(resolve => {
-        setTimeout(() => { resolve(); }, ms);
+import * as vscode from 'vscode';
+
+export async function delay(ms: number, token?: vscode.CancellationToken): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+        let cancellationListener: vscode.Disposable;
+
+        const timeout = setTimeout(() => {
+            cancellationListener?.dispose();
+            resolve();
+        }, ms);
+
+        if (token) {
+            cancellationListener = token.onCancellationRequested(() => {
+                cancellationListener.dispose();
+                clearTimeout(timeout);
+                reject();
+            });
+        }
     });
 }

--- a/src/utils/delay.ts
+++ b/src/utils/delay.ts
@@ -10,6 +10,7 @@ export async function delay(ms: number, token?: vscode.CancellationToken): Promi
         let cancellationListener: vscode.Disposable;
 
         const timeout = setTimeout(() => {
+            /* eslint-disable-next-line no-unused-expressions */
             cancellationListener?.dispose();
             resolve();
         }, ms);

--- a/src/utils/delay.ts
+++ b/src/utils/delay.ts
@@ -8,9 +8,3 @@ export async function delay(ms: number): Promise<void> {
         setTimeout(() => { resolve(); }, ms);
     });
 }
-
-export async function delayWithResult<T>(ms: number, result: T): Promise<T> {
-    return new Promise<T>(resolve => {
-        setTimeout(() => { resolve(result); }, ms);
-    });
-}

--- a/src/utils/delay.ts
+++ b/src/utils/delay.ts
@@ -8,3 +8,9 @@ export async function delay(ms: number): Promise<void> {
         setTimeout(() => { resolve(); }, ms);
     });
 }
+
+export async function delayWithResult<T>(ms: number, result: T): Promise<T> {
+    return new Promise<T>(resolve => {
+        setTimeout(() => { resolve(result); }, ms);
+    });
+}

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -9,7 +9,7 @@ import { Socket } from 'net';
 import { ext } from '../extensionVariables';
 import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
 import { cloneObject } from './cloneObject';
-import { delayWithResult } from './delay';
+import { delay } from './delay';
 import { isWindows } from './osUtils';
 import { execAsync } from './spawnAsync';
 
@@ -103,7 +103,7 @@ async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> 
     });
 
     // Unfortunately Socket.setTimeout() does not actually work when attempting to establish a connection, so we need to race
-    const result = await Promise.race([connectPromise, delayWithResult(1000, false)]);
+    const result = await Promise.race([connectPromise, delay(1000).then(() => false)]);
     authSock.end();
 
     return result;

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -3,6 +3,7 @@
  *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import { Socket } from 'net';
 import { DockerOptions } from 'dockerode';
 import Dockerode = require('dockerode');
 import { ext } from '../extensionVariables';
@@ -56,11 +57,13 @@ async function getDockerodeOptions(newEnv: NodeJS.ProcessEnv): Promise<DockerOpt
 
     try {
         if (newEnv.DOCKER_HOST &&
-            SSH_URL_REGEX.test(newEnv.DOCKER_HOST) &&
-            !newEnv.SSH_AUTH_SOCK) {
-            // If DOCKER_HOST is an SSH URL, we need to configure SSH_AUTH_SOCK for Dockerode
+            SSH_URL_REGEX.test(newEnv.DOCKER_HOST)) {
+            // If DOCKER_HOST is an SSH URL, we need to configure / validate SSH_AUTH_SOCK for Dockerode
             // Other than that, we use default settings, so return undefined
-            newEnv.SSH_AUTH_SOCK = await getSshAuthSock();
+            if (!await validateSshAuthSock(newEnv)) {
+                await ext.ui.showWarningMessage('In order to use an SSH DOCKER_HOST, you must configure an ssh-agent.', { learnMoreLink: 'https://aka.ms/AA7assy' });
+            }
+
             return undefined;
         } else if (!newEnv.DOCKER_HOST) {
             // If DOCKER_HOST is unset, try to get default Docker context--this helps support WSL
@@ -72,14 +75,31 @@ async function getDockerodeOptions(newEnv: NodeJS.ProcessEnv): Promise<DockerOpt
     return undefined;
 }
 
-async function getSshAuthSock(): Promise<string | undefined> {
-    if (isWindows()) {
-        return '\\\\.\\pipe\\openssh-ssh-agent';
-    } else {
+async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> {
+    if (!newEnv.SSH_AUTH_SOCK && isWindows()) {
+        // On Windows, we can use this one by default
+        newEnv.SSH_AUTH_SOCK = '\\\\.\\pipe\\openssh-ssh-agent';
+    } else if (!newEnv.SSH_AUTH_SOCK) {
         // On Mac and Linux, if SSH_AUTH_SOCK isn't set there's nothing we can do
         // Running ssh-agent would yield a new agent that doesn't have the needed keys
-        await ext.ui.showWarningMessage('In order to use an SSH DOCKER_HOST on OS X and Linux, you must configure an ssh-agent.');
+        return false;
     }
+
+    // Validate the pipe actually exists
+    return await new Promise<boolean>((resolve, reject) => {
+        const authPipe = new Socket();
+
+        authPipe.on('error', () => {
+            resolve(false);
+        });
+
+        authPipe.on('connect', () => {
+            authPipe.end();
+            resolve(true);
+        });
+
+        authPipe.connect(newEnv.SSH_AUTH_SOCK);
+    });
 }
 
 async function getDefaultDockerContext(): Promise<DockerOptions | undefined> {

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -9,6 +9,7 @@ import { Socket } from 'net';
 import { ext } from '../extensionVariables';
 import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
 import { cloneObject } from './cloneObject';
+import { delayWithResult } from './delay';
 import { isWindows } from './osUtils';
 import { execAsync } from './spawnAsync';
 
@@ -61,7 +62,9 @@ async function getDockerodeOptions(newEnv: NodeJS.ProcessEnv): Promise<DockerOpt
             // If DOCKER_HOST is an SSH URL, we need to configure / validate SSH_AUTH_SOCK for Dockerode
             // Other than that, we use default settings, so return undefined
             if (!await validateSshAuthSock(newEnv)) {
-                await ext.ui.showWarningMessage('In order to use an SSH DOCKER_HOST, you must configure an ssh-agent.', { learnMoreLink: 'https://aka.ms/AA7assy' });
+                // Don't wait
+                /* eslint-disable-next-line @typescript-eslint/no-floating-promises */
+                ext.ui.showWarningMessage('In order to use an SSH DOCKER_HOST, you must configure an ssh-agent.', { learnMoreLink: 'https://aka.ms/AA7assy' });
             }
 
             return undefined;
@@ -85,21 +88,25 @@ async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> 
         return false;
     }
 
-    // Validate the pipe actually exists
-    return await new Promise<boolean>((resolve, reject) => {
-        const authPipe = new Socket();
+    const authSock = new Socket();
 
-        authPipe.on('error', () => {
+    const connectPromise = new Promise<boolean>(resolve => {
+        authSock.on('error', (err) => {
             resolve(false);
         });
 
-        authPipe.on('connect', () => {
-            authPipe.end();
+        authSock.on('connect', () => {
             resolve(true);
         });
 
-        authPipe.connect(newEnv.SSH_AUTH_SOCK);
+        authSock.connect(newEnv.SSH_AUTH_SOCK);
     });
+
+    // Unfortunately Socket.setTimeout() does not actually work when attempting to establish a connection, so we need to race
+    const result = await Promise.race([connectPromise, delayWithResult(1000, false)]);
+    authSock.end();
+
+    return result;
 }
 
 async function getDefaultDockerContext(): Promise<DockerOptions | undefined> {

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -107,7 +107,11 @@ async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> 
     });
 
     // Unfortunately Socket.setTimeout() does not actually work when attempting to establish a connection, so we need to race
-    return await Promise.race([connectPromise, delay(1000, cts.token).then(() => false)]).finally(() => authSock.end());
+    return await Promise.race([connectPromise, delay(1000, cts.token).then(() => false)])
+        .finally(() => {
+            authSock.end();
+            cts.dispose();
+        });
 }
 
 async function getDefaultDockerContext(): Promise<DockerOptions | undefined> {

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -3,9 +3,9 @@
  *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { Socket } from 'net';
-import { DockerOptions } from 'dockerode';
 import Dockerode = require('dockerode');
+import { DockerOptions } from 'dockerode';
+import { Socket } from 'net';
 import { ext } from '../extensionVariables';
 import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
 import { cloneObject } from './cloneObject';

--- a/src/utils/refreshDockerode.ts
+++ b/src/utils/refreshDockerode.ts
@@ -6,6 +6,7 @@
 import Dockerode = require('dockerode');
 import { DockerOptions } from 'dockerode';
 import { Socket } from 'net';
+import { CancellationTokenSource } from 'vscode';
 import { ext } from '../extensionVariables';
 import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
 import { cloneObject } from './cloneObject';
@@ -89,13 +90,16 @@ async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> 
     }
 
     const authSock = new Socket();
+    const cts = new CancellationTokenSource();
 
     const connectPromise = new Promise<boolean>(resolve => {
         authSock.on('error', (err) => {
+            cts.cancel();
             resolve(false);
         });
 
         authSock.on('connect', () => {
+            cts.cancel();
             resolve(true);
         });
 
@@ -103,10 +107,7 @@ async function validateSshAuthSock(newEnv: NodeJS.ProcessEnv): Promise<boolean> 
     });
 
     // Unfortunately Socket.setTimeout() does not actually work when attempting to establish a connection, so we need to race
-    const result = await Promise.race([connectPromise, delay(1000).then(() => false)]);
-    authSock.end();
-
-    return result;
+    return await Promise.race([connectPromise, delay(1000, cts.token).then(() => false)]).finally(() => authSock.end());
 }
 
 async function getDefaultDockerContext(): Promise<DockerOptions | undefined> {

